### ISSUE TITLE
Add things on the fly while the server is running

### ISSUE THREAD

**Issue by tsathishkumar** [mozilla-iot/webthing-rust#3]
For the gateway's it would be nice to add the things on the fly as the gateway's discover the local things.

Currently the things are created before starting the server and I don't see a way to push more things once the server is started. Is there a way?

**Comment by mrstegeman** [comment]
What's your use case? This is not intended to be used as a gateway, but rather as a standalone web thing.

**Comment by mrstegeman** [comment]
The webthing libraries were intentionally built to serve as standalone web things. We had no need to implement another gateway, as we already [have one](https://github.com/mozilla-iot/gateway).

Could I instead suggest building an add-on for our gateway? I think that could be useful to a lot of people.

For reference, you can check out these articles:
* https://hacks.mozilla.org/2018/02/creating-an-add-on-for-the-project-things-gateway/
* https://hacks.mozilla.org/2018/02/making-a-clap-sensing-web-thing/

If you're particularly interested in Rust, it's possible to build an add-on in Rust. Here are some work-in-progress Rust add-ons:
* https://github.com/hobinjk/mqtt-adapter
* https://github.com/hobinjk/philips-hue-adapter-rs

**Comment by tsathishkumar** [comment]
I was thinking that it is easier to change the things from `Arc` to `Mutex` than writing a custom add-on. 

But if add-ons is the way to go then I'll take that path.

**Comment by mrstegeman** [comment]
I'd imagine writing an add-on will take you a lot less time than you suspect. 😄 

Please let us know if we can help at all! You can file issues, post on [Discourse](https://discourse.mozilla.org/c/iot), or find us in `#iot` on [irc.mozilla.org](https://wiki.mozilla.org/IRC).

**Comment by tsathishkumar** [comment]
But I would be simply forking this code and changing things from `Arc` to `Mutex` and then expose it as an add-on. I'm not sure how that would take less time :/

**Comment by mrstegeman** [comment]
Add-ons are not proper web things, so this code would not be what you want. They talk to the gateway through an IPC mechanism.

But yes, you're also welcome to fork this to do what you need. If you end up using this library, you can still add it to the gateway by URL.

I guess I'm still confused why you need to add things dynamically. Would they not all be available at startup?

**Comment by tsathishkumar** [comment]
Basically you can add a new sensor to the mysensors network and it will be automatically discovered by the My-Seonsor's Gateway. In order to publish the newly added sensor, I should be able to push it to the things vector whenever the gateway discovers the thing.

If I can' add things dynamically, then I need to add things, by inserting an entry to database, and then restart the server.

**Comment by tsathishkumar** [comment]
Hmmm.. Thinking bit more, maybe we can start the server with either Single thing mode or Multiple thing mode? I know it's bit verbose, but will be more explicit about the functionality. What do you think?

**Comment by tsathishkumar** [comment]
If it makes sense, I can work on the PR ;)

**Comment by mrstegeman** [comment]
I still just think this makes more sense as a gateway add-on. It seems very similar to Zigbee/Z-Wave, which we also have add-ons for.

However, feel free to submit a PR. I'll check it out and discuss with the team to see if that's a feature we want to maintain across all of the webthing libraries. If we decide not to merge, it will still be useful to you.

**Comment by tsathishkumar** [comment]
Great! To me Gateway as library/crate makes more sense than the existing (webthing) gateway - as executable server. It seems too coupled. Just my opinion.

**Comment by tsathishkumar** [comment]
Looking more at the code, it seems not possible to avoid server restart even with single/multiple thing mode. So I'm gonna stick to the restart approach for now.

However single/multiple thing mode seems a nice to have feature. I'll work on that.
